packages/cardpay-reward-scheduler/reward_scheduler/reward_program.py

```diff
@@ -204,22 +212,43 @@ def run(self, payment_cycle, rule):
         )
         return job
 
-    def run_all_payment_cycles(self) -> None:
+    def run_rule(self, rule) -> None:
+        processable_payment_cycles = (
+            self.get_all_payment_cycles(rule) - self.processed_cycles
+        )
+        for payment_cycle in sorted(processable_payment_cycles):
+            if rule["core"].get("rollover"):
+                previous_cycle = payment_cycle - rule["core"]["payment_cycle_length"]
+                if payment_cycle == rule["core"]["start_block"]:
+                    # There can't be any previous ones for the first cycle
+                    self.run_payment_cycle(payment_cycle, rule)
+                elif previous_cycle in self.processed_cycles:
+                    # Only run if the previous cycle has been processed
+                    # because the output of the previous cycle is needed
+                    self.run_payment_cycle(payment_cycle, rule, previous_cycle)
+                else:
+                    # There's nothing to do here as the previous cycle hasn't been processed
+                    pass
+
+            else:
+                self.run_payment_cycle(payment_cycle, rule)
+
+    def run_all_rules(self) -> None:
         """
         Run all rules for all payment cycles that can be processed at this time
         """
         logging.info(f"Running all for {self.reward_program_id}")
         if self.is_locked():
             logging.info(f"Reward program {self.reward_program_id} is locked, skipping")
             return
+
+        self.update_processed()
```

Comment:
Question: Given we `load_submitted_from_subgraph_extracction`, is it possible we can completely remove `update_processed` function which relies on subgraph??

Reply:
Yeah we could - only impact is latency. Subgraph gives us the data pretty much immediately. 

I'll have a check to see how much it would add because removing a dependency on subgraph completely sounds nice.